### ISSUE TITLE
How to count?

### ISSUE THREAD

**Issue by labibramadhan** [IRCraziestTaxi/typeorm-linq-repository#56]
How to get only count of the results by using `getAll` method?

On the one side i want to fetch entity results paged (using skip & take), and on the another side i want to fetch total results without paged (without using skip & take). The entity results are going to be shown for table paged data, and the total results are going to be used on the pagination total count for showing total available pages.

Or maybe there is a feature to clone query (IQuery)?

**Comment by labibramadhan** [comment]
Temporary solution:

```typescript
const q = userRepository.getAll();
const users = await q.skip(0).take(10);
const totalUsers = (await q.skip(0).take(10000000).select(u => u.id)).length;
```

But the `.select(u => u.id)` doesn't work, the returned entities are the full object rather than only returning id. I saw that [this line](https://github.com/IRCraziestTaxi/typeorm-linq-repository/blob/53bf369e53136179b4d5b2f835781695784cdede/src/query/Query.ts#L296), doesn't do nothing to the `_queryParts`, it doesn't push any kind of `QueryBuilderPart`, or maybe `buildQuery ` or `compileQueryParts` should call `query.query.select` on the end based on `query.selected` or `_selectedProperty` (if value is not undefined), and i think it would be great if `_selectedProperty` supports `string[]` rather than only `string` for selecting multiple columns

**Comment by IRCraziestTaxi** [comment]
Thanks for bringing this up! Along with a lot of other things I have been thinking about that this lib is lacking (one major one is grouping), I have been thinking about the lack of ability to count lately, especially when you opened this issue.

As soon as the issue was opened, I got an idea of how to implement this and just published that fix in version `1.0.0-alpha.17`.

Here is the rundown of what I did: I wanted to have a reusable query reminiscent of Entity Framework's `IQueryable`, so I implemented `LinqRepository.query`, which returns an `IQuery` that can be reused with the following methods: `count()`, `getMany()`, and `getOne()`.

For example:

```ts
// Previously:
this._userRepository
    .getAll()
    .where(u => u.active)
    .isTrue();
// Query is now "dead" and not reusable.

// Now:
let activeUserQuery = this._userRepository
    .query
    .where(u => u.active)
    .isTrue();

// Count all active users.
const activeUserCount = await activeUserQuery.count();

// Apply constraints for paging, etc.
activeUserQuery = activeUserQuery
    .skip(skip)
    .take(take);

const activeUserPageResults = activeUserQuery.getMany();
```

For the sake of backwards compatibility, the "one-off" methods of querying via `LinqRepository.get*` are still supported, although `count()` can only be used via `LinqRepository.query`.

In light of that, I'm going to close the PR linked to this issue since this implementation addresses both ability to count as well as the "reusable query" concern.

**Comment by IRCraziestTaxi** [comment]
Actually, I was thinking about my changes after I published them and I realized that I was way over-thinking the problem. Adding a `count` method that clones the QueryBuilder was, in fact, sufficient to solve the problem, since the only reason one would ever need to "re-use" a query is for counting. My changes just added unnecessary complexity to the code and confusion to the usage of queries.

I'm going to unpublish the version I published last night, revert my changes, and go with yours instead.
